### ISSUE TITLE
3.1.3 ResultSetDecorator isEmpty deletes 1 result

### ISSUE THREAD

**Issue by Graziel** [cakephp/cakephp#7649]
ive this piece of code

```PHP

$this->loadModel('BusinessesCategoriesAll');

            $categories_query = $this->BusinessesCategoriesAll->find()
                ->select(['id', 'name' => "name_$lang"])
                ->where(["name_$lang LIKE" => $data['term'] . '%']);

            if (!empty($searches['cat'])) {
                $categories_query->where(['id NOT IN' => $searches['cat']]);
            }

            $categories = $categories_query
                ->limit(5)
                ->formatResults(function (\Cake\Datasource\ResultSetInterface $resultzs) {
                    return $resultzs->map(function ($row) {
                        $row['id'] = 'cat:' . $row['id'];
                        $row['__group_id'] = 3;
                        return $row;
                    });
                })
                ->all();

            pr($categories); // comment it

            if (!$categories->isEmpty()) {
                pr($categories);
                $results[] = [
                    '__group_id' => 3,
                    'text' => __('Categories'),
                    'children' => $categories
                ];
            }
```

when it looks like it gives correct amount of results

```
Cake\Datasource\ResultSetDecorator Object
(
    [count] => 5
)
Cake\Datasource\ResultSetDecorator Object
(
    [count] => 5
)
```

but when i comment the 1st pr

```
Cake\Datasource\ResultSetDecorator Object
(
    [count] => 4
)
```

**Comment by markstory** [comment]
Counts in collection objects are not very reliable. See #7565

**Comment by Graziel** [comment]
changed pr($categories) to pr($categories->toArray())

```
Array
(
    [0] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:37
            [name] => Basketball
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 37
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [1] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:77
            [name] => Bakery
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 77
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [2] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:162
            [name] => Bank
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 162
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [3] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:328
            [name] => Baltic States
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 328
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [4] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:348
            [name] => Basque
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 348
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

)
Array
(
    [0] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:37
            [name] => Basketball
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 37
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [1] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:77
            [name] => Bakery
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 77
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [2] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:162
            [name] => Bank
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 162
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [3] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:328
            [name] => Baltic States
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 328
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [4] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:348
            [name] => Basque
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 348
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

)
```

```
Array
(
    [0] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:37
            [name] => Basketball
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 37
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [1] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:77
            [name] => Bakery
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 77
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [3] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:328
            [name] => Baltic States
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 328
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

    [4] => App\Model\Entity\BusinessesCategoriesAll Object
        (
            [id] => cat:348
            [name] => Basque
            [__group_id] => 3
            [[new]] => 
            [[accessible]] => Array
                (
                    [*] => 1
                )

            [[dirty]] => Array
                (
                    [id] => 1
                    [__group_id] => 1
                )

            [[original]] => Array
                (
                    [id] => 348
                )

            [[virtual]] => Array
                (
                )

            [[errors]] => Array
                (
                )

            [[repository]] => BusinessesCategoriesAll
        )

)
```

note there is no [2] on second pass

**Comment by markstory** [comment]
If you don't apply a map() function do the results change?

**Comment by Graziel** [comment]
map() no change, but when i remove whole ->formatResults(), it gives correct results

**Comment by markstory** [comment]
Interesting, I can see if I can reproduce locally and dig deeper.

**Comment by markstory** [comment]
I was able to reduce this to the following (which currently fails)

```php
        $table = TableRegistry::get('Comments');
        $query = $table->find()
            ->limit(5)
            ->formatResults(function ($results) {
                return $results;
            });
        $res = $query->all();
        $res->isEmpty();
        $this->assertCount(5, $res->toArray());
```

If either the `limit()` or `formatResults()` calls are removed the test passes. I've not had a chance to dig deeper just yet, but it seems like interaction of the various stacked iterators is causing issues.

**Comment by markstory** [comment]
Fixed with #7664
